In [13]:
import pandas as pd

In [14]:
df=pd.read_csv("/kaggle/input/datasets/rajpratapsinghsengar/clean-dataset/dataset.csv")

In [15]:
df.head()

,Unnamed: 0,age,gender,marital_status,education,monthly_salary,employment_type,years_of_employment,company_type,house_type,...,existing_loans,current_emi_amount,credit_score,bank_balance,emergency_fund,emi_scenario,requested_amount,requested_tenure,emi_eligibility,max_monthly_emi
0,0,38.0,Female,Married,Professional,82600.0,Private,0.9,Mid-size,Rented,...,Yes,23700.0,660.0,303200.0,70200.0,Personal Loan EMI,850000.0,15,Not_Eligible,500.0
1,1,38.0,Female,Married,Graduate,21500.0,Private,7.0,MNC,Family,...,Yes,4100.0,714.0,92500.0,26900.0,E-commerce Shopping EMI,128000.0,19,Not_Eligible,700.0
2,2,38.0,Male,Married,Professional,86100.0,Private,5.8,Startup,Own,...,No,0.0,650.0,672100.0,324200.0,Education EMI,306000.0,16,Eligible,27775.0
3,3,58.0,Female,Married,High School,66800.0,Private,2.2,Mid-size,Own,...,No,0.0,685.0,440900.0,178100.0,Vehicle EMI,304000.0,83,Eligible,16170.0
4,4,48.0,Female,Married,Professional,57300.0,Private,3.4,Mid-size,Family,...,No,0.0,770.0,97300.0,28200.0,Home Appliances EMI,252000.0,7,Not_Eligible,500.0


In [16]:
df.shape

(404800, 27)

In [17]:
dataset=df.copy()

In [18]:
dataset.shape

(404800, 27)

In [19]:
X = dataset.drop(
    columns=["emi_eligibility", "max_monthly_emi","Unnamed: 0"],
    errors="ignore"
)

y = dataset["emi_eligibility"]

In [20]:
X.shape

(404800, 24)

In [21]:
y.shape

(404800,)

In [22]:
X.columns

Index(['age', 'gender', 'marital_status', 'education', 'monthly_salary',
       'employment_type', 'years_of_employment', 'company_type', 'house_type',
       'monthly_rent', 'family_size', 'school_fees', 'college_fees',
       'travel_expenses', 'groceries_utilities', 'other_monthly_expenses',
       'existing_loans', 'current_emi_amount', 'credit_score', 'bank_balance',
       'emergency_fund', 'emi_scenario', 'requested_amount',
       'requested_tenure'],
      dtype='object')

In [23]:
y.head()

0    Not_Eligible
1    Not_Eligible
2        Eligible
3        Eligible
4    Not_Eligible
Name: emi_eligibility, dtype: object

In [24]:
print(y.value_counts())

print("\nClass Percentage:")
print(y.value_counts(normalize=True) * 100)

emi_eligibility
Not_Eligible    312868
Eligible         74444
High_Risk        17488
Name: count, dtype: int64

Class Percentage:
emi_eligibility
Not_Eligible    77.289526
Eligible        18.390316
High_Risk        4.320158
Name: proportion, dtype: float64


In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    stratify=y,
    random_state=42
)

In [26]:
numerical_with_nan = [
    feature
    for feature in X_train.columns
    if X_train[feature].isnull().sum() > 0
    and X_train[feature].dtype != "object"
]

print(numerical_with_nan)

['age', 'monthly_salary', 'monthly_rent', 'credit_score', 'bank_balance', 'emergency_fund']


In [27]:
for feature in numerical_with_nan:
    print(X_train[feature].dtypes)

float64
float64
float64
float64
float64
float64


In [28]:
from sklearn.impute import SimpleImputer

num_imputer= SimpleImputer(strategy="median")

X_train[numerical_with_nan] = num_imputer.fit_transform(
    X_train[numerical_with_nan]
)

X_test[numerical_with_nan] = num_imputer.transform(
    X_test[numerical_with_nan]
)

In [29]:
X_train[numerical_with_nan].isnull().sum()

age               0
monthly_salary    0
monthly_rent      0
credit_score      0
bank_balance      0
emergency_fund    0
dtype: int64

In [30]:
X_test[numerical_with_nan].isnull().sum()

age               0
monthly_salary    0
monthly_rent      0
credit_score      0
bank_balance      0
emergency_fund    0
dtype: int64

In [31]:
cat_imputer = SimpleImputer(strategy="most_frequent")

X_train[["education"]] = cat_imputer.fit_transform(
    X_train[["education"]]
)

X_test[["education"]] = cat_imputer.transform(
    X_test[["education"]]
)

In [32]:
X_train.isnull().sum()

age                       0
gender                    0
marital_status            0
education                 0
monthly_salary            0
employment_type           0
years_of_employment       0
company_type              0
house_type                0
monthly_rent              0
family_size               0
school_fees               0
college_fees              0
travel_expenses           0
groceries_utilities       0
other_monthly_expenses    0
existing_loans            0
current_emi_amount        0
credit_score              0
bank_balance              0
emergency_fund            0
emi_scenario              0
requested_amount          0
requested_tenure          0
dtype: int64

In [33]:
X_test.isnull().sum()

age                       0
gender                    0
marital_status            0
education                 0
monthly_salary            0
employment_type           0
years_of_employment       0
company_type              0
house_type                0
monthly_rent              0
family_size               0
school_fees               0
college_fees              0
travel_expenses           0
groceries_utilities       0
other_monthly_expenses    0
existing_loans            0
current_emi_amount        0
credit_score              0
bank_balance              0
emergency_fund            0
emi_scenario              0
requested_amount          0
requested_tenure          0
dtype: int64

In [34]:
yeo_johnson_features = [
    'age',
    'monthly_salary',
    'years_of_employment',
    'monthly_rent',
    'college_fees',
    'travel_expenses',
    'groceries_utilities',
    'other_monthly_expenses',
    'current_emi_amount',
    'credit_score',
    'bank_balance',
    'emergency_fund',
    'requested_amount',
    'requested_tenure'
]

In [35]:
import numpy as np

In [36]:
X_train["school_fees"] = np.sqrt(
    X_train["school_fees"]
)

X_test["school_fees"] = np.sqrt(
    X_test["school_fees"]
)

In [37]:
from sklearn.preprocessing import PowerTransformer

pt_cls = PowerTransformer(
    method="yeo-johnson",
    standardize=False
)

X_train[yeo_johnson_features] = pt_cls.fit_transform(
    X_train[yeo_johnson_features]
)

X_test[yeo_johnson_features] = pt_cls.transform(
    X_test[yeo_johnson_features]
)
print("done")

done


In [38]:
categorical_features = [
    feature
    for feature in X_train.columns
    if X_train[feature].dtype == "object"
]

print(categorical_features)

['gender', 'marital_status', 'education', 'employment_type', 'company_type', 'house_type', 'existing_loans', 'emi_scenario']


In [39]:
from sklearn.preprocessing import OneHotEncoder

encoder_cls = OneHotEncoder(
    drop="first",
    handle_unknown="ignore",
    sparse_output=False
)
print("done")

done


In [40]:
X_train_encoded = encoder_cls.fit_transform(
    X_train[categorical_features]
)

X_test_encoded = encoder_cls.transform(
    X_test[categorical_features]
)
print("done")

done


In [41]:
encoded_columns = encoder_cls.get_feature_names_out(
    categorical_features
)

X_train_encoded = pd.DataFrame(
    X_train_encoded,
    columns=encoded_columns,
    index=X_train.index
)

X_test_encoded = pd.DataFrame(
    X_test_encoded,
    columns=encoded_columns,
    index=X_test.index
)
print("done")

done


In [42]:
X_train = X_train.drop(
    columns=categorical_features
)

X_test = X_test.drop(
    columns=categorical_features
)

In [43]:
X_train.shape

(323840, 16)

In [44]:
X_train_final = pd.concat(
    [X_train, X_train_encoded],
    axis=1
)

X_test_final = pd.concat(
    [X_test, X_test_encoded],
    axis=1
)

In [45]:
print(X_train_final.shape)
print(X_test_final.shape)

(323840, 34)
(80960, 34)


In [46]:
import pandas as pd
import numpy as np

from sklearn.feature_selection import VarianceThreshold
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.pipeline import Pipeline

In [47]:
X_train_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 323840 entries, 134637 to 100297
Data columns (total 34 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   age                               323840 non-null  float64
 1   monthly_salary                    323840 non-null  float64
 2   years_of_employment               323840 non-null  float64
 3   monthly_rent                      323840 non-null  float64
 4   family_size                       323840 non-null  int64  
 5   school_fees                       323840 non-null  float64
 6   college_fees                      323840 non-null  float64
 7   travel_expenses                   323840 non-null  float64
 8   groceries_utilities               323840 non-null  float64
 9   other_monthly_expenses            323840 non-null  float64
 10  current_emi_amount                323840 non-null  float64
 11  credit_score                      323840 non-null  f

In [48]:
variance_selector = VarianceThreshold(threshold=0.01)

X_train_var = variance_selector.fit_transform(X_train_final)
X_test_var = variance_selector.transform(X_test_final)

variance_features = X_train_final.columns[
    variance_selector.get_support()
]

X_train_var = pd.DataFrame(
    X_train_var,
    columns=variance_features,
    index=X_train_final.index
)

X_test_var = pd.DataFrame(
    X_test_var,
    columns=variance_features,
    index=X_test_final.index
)

print("Features before variance:", X_train_final.shape[1])
print("Features after variance:", X_train_var.shape[1])

Features before variance: 34
Features after variance: 33


In [49]:
X_train_final.columns

Index(['age', 'monthly_salary', 'years_of_employment', 'monthly_rent',
       'family_size', 'school_fees', 'college_fees', 'travel_expenses',
       'groceries_utilities', 'other_monthly_expenses', 'current_emi_amount',
       'credit_score', 'bank_balance', 'emergency_fund', 'requested_amount',
       'requested_tenure', 'gender_Male', 'marital_status_Single',
       'education_High School', 'education_Post Graduate',
       'education_Professional', 'employment_type_Private',
       'employment_type_Self-employed', 'company_type_MNC',
       'company_type_Mid-size', 'company_type_Small', 'company_type_Startup',
       'house_type_Own', 'house_type_Rented', 'existing_loans_Yes',
       'emi_scenario_Education EMI', 'emi_scenario_Home Appliances EMI',
       'emi_scenario_Personal Loan EMI', 'emi_scenario_Vehicle EMI'],
      dtype='object')

In [50]:
corr_matrix = X_train_var.corr().abs()

upper_triangle = corr_matrix.where(
    np.triu(
        np.ones(corr_matrix.shape),
        k=1
    ).astype(bool)
)

correlated_features = [
    column
    for column in upper_triangle.columns
    if any(upper_triangle[column] > 0.90)
]

print("Highly correlated features:", len(correlated_features))
print(correlated_features)

Highly correlated features: 3
['emergency_fund', 'house_type_Rented', 'existing_loans_Yes']


In [51]:
X_train_corr = X_train_var.drop(
    columns=correlated_features
)

X_test_corr = X_test_var.drop(
    columns=correlated_features
)

print("Features after correlation:", X_train_corr.shape[1])

Features after correlation: 30


In [52]:
rf = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1,
    class_weight="balanced"
)

rf.fit(X_train_corr, y_train)

RandomForestClassifier(class_weight='balanced', n_estimators=200, n_jobs=-1,
                       random_state=42)

In [53]:
feature_importance = pd.DataFrame({
    "feature": X_train_corr.columns,
    "importance": rf.feature_importances_
})

feature_importance = feature_importance.sort_values(
    by="importance",
    ascending=False
)

print(feature_importance)

                             feature  importance
12                  requested_amount    0.196592
0                     monthly_salary    0.100480
13                  requested_tenure    0.094192
9                 current_emi_amount    0.062040
11                      bank_balance    0.059883
7                groceries_utilities    0.058431
10                      credit_score    0.053310
6                    travel_expenses    0.049398
8             other_monthly_expenses    0.046211
4                        school_fees    0.045341
2                       monthly_rent    0.045275
5                       college_fees    0.040849
1                years_of_employment    0.039743
3                        family_size    0.017127
25                    house_type_Own    0.008142
14                       gender_Male    0.007178
19           employment_type_Private    0.006705
21                  company_type_MNC    0.006355
22             company_type_Mid-size    0.006315
17           educati

In [54]:
ranked_features = feature_importance["feature"].tolist()

print(ranked_features)

['requested_amount', 'monthly_salary', 'requested_tenure', 'current_emi_amount', 'bank_balance', 'groceries_utilities', 'credit_score', 'travel_expenses', 'other_monthly_expenses', 'school_fees', 'monthly_rent', 'college_fees', 'years_of_employment', 'family_size', 'house_type_Own', 'gender_Male', 'employment_type_Private', 'company_type_MNC', 'company_type_Mid-size', 'education_Post Graduate', 'emi_scenario_Vehicle EMI', 'emi_scenario_Home Appliances EMI', 'marital_status_Single', 'emi_scenario_Personal Loan EMI', 'education_Professional', 'emi_scenario_Education EMI', 'company_type_Startup', 'education_High School', 'employment_type_Self-employed', 'company_type_Small']


In [55]:
k_values = [5, 10, 15, 20, 25, 30]

k_values = [
    k for k in k_values
    if k <= len(ranked_features)
]

print("K values:", k_values)

K values: [5, 10, 15, 20, 25, 30]


In [56]:
cv = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

In [57]:
results = []

for k in k_values:

    selected_features = ranked_features[:k]

    X_k = X_train_corr[selected_features]

    model = RandomForestClassifier(
        n_estimators=200,
        random_state=42,
        n_jobs=-1
    )

    scores = cross_val_score(
        model,
        X_k,
        y_train,
        cv=cv,
        scoring="f1_macro",
        n_jobs=-1
    )

    results.append({
        "K": k,
        "Mean_Macro_F1": scores.mean(),
        "Std_Macro_F1": scores.std()
    })

results_df = pd.DataFrame(results)

print(results_df)

/usr/local/lib/python3.12/dist-packages/joblib/externals/loky/process_executor.py:782: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/joblib/externals/loky/process_executor.py:782: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


    K  Mean_Macro_F1  Std_Macro_F1
0   5       0.549429      0.001659
1  10       0.569627      0.000528
2  15       0.624532      0.000835
3  20       0.622956      0.000648
4  25       0.622573      0.000555
5  30       0.619121      0.000791


In [58]:
best_k = int(
    results_df.loc[
        results_df["Mean_Macro_F1"].idxmax(),
        "K"
    ]
)

print("Best K:", best_k)

Best K: 15


In [59]:
best_features = ranked_features[:best_k]

print("Selected features:")

for i, feature in enumerate(best_features, 1):
    print(i, feature)

Selected features:
1 requested_amount
2 monthly_salary
3 requested_tenure
4 current_emi_amount
5 bank_balance
6 groceries_utilities
7 credit_score
8 travel_expenses
9 other_monthly_expenses
10 school_fees
11 monthly_rent
12 college_fees
13 years_of_employment
14 family_size
15 house_type_Own


In [61]:
best_k = 15

best_features = ranked_features[:best_k]

X_train_final = X_train_corr[best_features]
X_test_final = X_test_corr[best_features]

print("Best K:", best_k)
print("X_train_final:", X_train_final.shape)
print("X_test_final :", X_test_final.shape)

print("\nSelected Features:")
print(best_features)

Best K: 15
X_train_final: (323840, 15)
X_test_final : (80960, 15)

Selected Features:
['requested_amount', 'monthly_salary', 'requested_tenure', 'current_emi_amount', 'bank_balance', 'groceries_utilities', 'credit_score', 'travel_expenses', 'other_monthly_expenses', 'school_fees', 'monthly_rent', 'college_fees', 'years_of_employment', 'family_size', 'house_type_Own']


In [62]:
import pandas as pd

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from xgboost import XGBClassifier

In [63]:
from sklearn.model_selection import train_test_split

X_train_base, X_val, y_train_base, y_val = train_test_split(
    X_train_final,
    y_train,
    test_size=0.20,
    stratify=y_train,
    random_state=42
)

print("X_train_base:", X_train_base.shape)
print("X_val:", X_val.shape)
print("X_test_final:", X_test_final.shape)

print("y_train_base:", y_train_base.shape)
print("y_val:", y_val.shape)
print("y_test:", y_test.shape)

X_train_base: (259072, 15)
X_val: (64768, 15)
X_test_final: (80960, 15)
y_train_base: (259072,)
y_val: (64768,)
y_test: (80960,)


In [64]:
print("Training distribution:")
print(y_train_base.value_counts(normalize=True))

print("\nValidation distribution:")
print(y_val.value_counts(normalize=True))

print("\nTest distribution:")
print(y_test.value_counts(normalize=True))

Training distribution:
emi_eligibility
Not_Eligible    0.772893
Eligible        0.183903
High_Risk       0.043204
Name: proportion, dtype: float64

Validation distribution:
emi_eligibility
Not_Eligible    0.772897
Eligible        0.183903
High_Risk       0.043200
Name: proportion, dtype: float64

Test distribution:
emi_eligibility
Not_Eligible    0.772900
Eligible        0.183906
High_Risk       0.043194
Name: proportion, dtype: float64


In [67]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()

y_train_base_encoded = label_encoder.fit_transform(y_train_base)
y_val_encoded = label_encoder.transform(y_val)

print("Classes:", label_encoder.classes_)

Classes: ['Eligible' 'High_Risk' 'Not_Eligible']


In [70]:
models = {

    "Random Forest": RandomForestClassifier(
        n_estimators=200,
        random_state=42,
        n_jobs=-1
    ),

    "LightGBM": LGBMClassifier(
        random_state=42,
        n_jobs=-1,
        verbosity=-1
    ),

    "CatBoost": CatBoostClassifier(
        iterations=200,
        random_state=42,
        verbose=False
    ),

    "XGBoost": XGBClassifier(
        n_estimators=200,
        random_state=42,
        n_jobs=-1,
        eval_metric="mlogloss"
    )
}

In [71]:
baseline_results = []

for name, model in models.items():

    print(f"Training {name}...")

    model.fit(
        X_train_base,
        y_train_base_encoded
    )

    y_val_pred = model.predict(X_val)

    baseline_results.append({
        "Model": name,
        "Accuracy": accuracy_score(
            y_val_encoded,
            y_val_pred
        ),
        "Balanced Accuracy": balanced_accuracy_score(
            y_val_encoded,
            y_val_pred
        ),
        "Macro Precision": precision_score(
            y_val_encoded,
            y_val_pred,
            average="macro",
            zero_division=0
        ),
        "Macro Recall": recall_score(
            y_val_encoded,
            y_val_pred,
            average="macro",
            zero_division=0
        ),
        "Macro F1": f1_score(
            y_val_encoded,
            y_val_pred,
            average="macro",
            zero_division=0
        )
    })

baseline_results_df = pd.DataFrame(baseline_results)

baseline_results_df.sort_values(
    by="Macro F1",
    ascending=False
)

Training Random Forest...
Training LightGBM...
Training CatBoost...
Training XGBoost...


,Model,Accuracy,Balanced Accuracy,Macro Precision,Macro Recall,Macro F1
3,XGBoost,0.953418,0.715139,0.835890,0.715139,0.735539
2,CatBoost,0.952168,0.691808,0.842131,0.691808,0.701555
1,LightGBM,0.937176,0.644255,0.753862,0.644255,0.637759
0,Random Forest,0.926383,0.625347,0.873735,0.625347,0.625003


In [73]:
from sklearn.metrics import classification_report, confusion_matrix

xgb_model = models["XGBoost"]

y_val_pred_xgb = xgb_model.predict(X_val)

print("Classification Report:")
print(
    classification_report(
        y_val_encoded,
        y_val_pred_xgb,
        target_names=label_encoder.classes_,
        zero_division=0
    )
)

print("\nConfusion Matrix:")
print(
    confusion_matrix(
        y_val_encoded,
        y_val_pred_xgb
    )
)

Classification Report:
              precision    recall  f1-score   support

    Eligible       0.91      0.96      0.94     11911
   High_Risk       0.63      0.19      0.29      2798
Not_Eligible       0.97      0.99      0.98     50059

    accuracy                           0.95     64768
   macro avg       0.84      0.72      0.74     64768
weighted avg       0.94      0.95      0.94     64768


Confusion Matrix:
[[11494   176   241]
 [  925   523  1350]
 [  190   135 49734]]


In [74]:
import numpy as np
from sklearn.utils.class_weight import compute_sample_weight

In [75]:
sample_weights = compute_sample_weight(
    class_weight="balanced",
    y=y_train_base_encoded
)

print("Sample weights shape:", sample_weights.shape)
print("Minimum weight:", sample_weights.min())
print("Maximum weight:", sample_weights.max())

Sample weights shape: (259072,)
Minimum weight: 0.4312799127691629
Maximum weight: 7.715298251883618


In [76]:
from xgboost import XGBClassifier

xgb_balanced = XGBClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1,
    eval_metric="mlogloss"
)

In [77]:
xgb_balanced.fit(
    X_train_base,
    y_train_base_encoded,
    sample_weight=sample_weights
)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=200, n_jobs=-1,
              num_parallel_tree=None, ...)

In [78]:
y_val_pred_balanced = xgb_balanced.predict(
    X_val
)

In [79]:
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

balanced_results = {
    "Accuracy": accuracy_score(
        y_val_encoded,
        y_val_pred_balanced
    ),

    "Balanced Accuracy": balanced_accuracy_score(
        y_val_encoded,
        y_val_pred_balanced
    ),

    "Macro Precision": precision_score(
        y_val_encoded,
        y_val_pred_balanced,
        average="macro",
        zero_division=0
    ),

    "Macro Recall": recall_score(
        y_val_encoded,
        y_val_pred_balanced,
        average="macro",
        zero_division=0
    ),

    "Macro F1": f1_score(
        y_val_encoded,
        y_val_pred_balanced,
        average="macro",
        zero_division=0
    )
}

balanced_results

{'Accuracy': 0.9179996294466403,
 'Balanced Accuracy': np.float64(0.8671495909885257),
 'Macro Precision': 0.7550951901749033,
 'Macro Recall': 0.8671495909885257,
 'Macro F1': 0.7809058554458961}

In [80]:
from sklearn.metrics import classification_report

print(
    classification_report(
        y_val_encoded,
        y_val_pred_balanced,
        target_names=label_encoder.classes_,
        zero_division=0
    )
)

              precision    recall  f1-score   support

    Eligible       0.94      0.89      0.91     11911
   High_Risk       0.33      0.78      0.47      2798
Not_Eligible       0.99      0.93      0.96     50059

    accuracy                           0.92     64768
   macro avg       0.76      0.87      0.78     64768
weighted avg       0.96      0.92      0.93     64768



In [81]:
print(confusion_matrix(
    y_val_encoded,
    y_val_pred_balanced
))

[[10615  1229    67]
 [  444  2177   177]
 [  266  3128 46665]]


In [82]:
from imblearn.over_sampling import SMOTE

In [83]:
print("Before SMOTE:")
print(y_train_base.value_counts())

print("\nPercentage:")
print(
    y_train_base.value_counts(normalize=True) * 100
)

Before SMOTE:
emi_eligibility
Not_Eligible    200235
Eligible         47644
High_Risk        11193
Name: count, dtype: int64

Percentage:
emi_eligibility
Not_Eligible    77.289325
Eligible        18.390254
High_Risk        4.320421
Name: proportion, dtype: float64


In [84]:
smote = SMOTE(
    random_state=42
)

X_train_smote, y_train_smote = smote.fit_resample(
    X_train_base,
    y_train_base_encoded
)

In [85]:
print("After SMOTE:")
print(
    pd.Series(y_train_smote).value_counts()
)

After SMOTE:
2    200235
1    200235
0    200235
Name: count, dtype: int64


In [86]:
print("Before SMOTE:", X_train_base.shape)
print("After SMOTE :", X_train_smote.shape)

print("Original y:", y_train_base_encoded.shape)
print("SMOTE y   :", y_train_smote.shape)

Before SMOTE: (259072, 15)
After SMOTE : (600705, 15)
Original y: (259072,)
SMOTE y   : (600705,)


In [87]:
from xgboost import XGBClassifier

xgb_smote = XGBClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1,
    eval_metric="mlogloss"
)

In [88]:
xgb_smote.fit(
    X_train_smote,
    y_train_smote
)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='mlogloss',
              feature_types=None, feature_weights=None, gamma=None,
              grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=200, n_jobs=-1,
              num_parallel_tree=None, ...)

In [89]:
y_val_pred_smote = xgb_smote.predict(
    X_val
)

In [90]:
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

smote_results = {
    "Accuracy": accuracy_score(
        y_val_encoded,
        y_val_pred_smote
    ),

    "Balanced Accuracy": balanced_accuracy_score(
        y_val_encoded,
        y_val_pred_smote
    ),

    "Macro Precision": precision_score(
        y_val_encoded,
        y_val_pred_smote,
        average="macro",
        zero_division=0
    ),

    "Macro Recall": recall_score(
        y_val_encoded,
        y_val_pred_smote,
        average="macro",
        zero_division=0
    ),

    "Macro F1": f1_score(
        y_val_encoded,
        y_val_pred_smote,
        average="macro",
        zero_division=0
    )
}

smote_results

{'Accuracy': 0.9505002470355731,
 'Balanced Accuracy': np.float64(0.7767124691735869),
 'Macro Precision': 0.8038439047832875,
 'Macro Recall': 0.7767124691735869,
 'Macro F1': 0.7886119057126502}

In [91]:
from sklearn.metrics import classification_report

print(
    classification_report(
        y_val_encoded,
        y_val_pred_smote,
        target_names=label_encoder.classes_,
        zero_division=0
    )
)

              precision    recall  f1-score   support

    Eligible       0.93      0.94      0.93     11911
   High_Risk       0.51      0.41      0.45      2798
Not_Eligible       0.98      0.98      0.98     50059

    accuracy                           0.95     64768
   macro avg       0.80      0.78      0.79     64768
weighted avg       0.95      0.95      0.95     64768



In [92]:
comparison = pd.DataFrame([
    {
        "Approach": "Baseline XGBoost",
        "Accuracy": 0.953418,
        "Balanced Accuracy": 0.715139,
        "Macro F1": 0.735539,
        "High_Risk Recall": 0.19,
        "High_Risk F1": 0.29
    },
    {
        "Approach": "Balanced Sample Weight XGBoost",
        "Accuracy": 0.9179996,
        "Balanced Accuracy": 0.8671496,
        "Macro F1": 0.7809059,
        "High_Risk Recall": 0.78,
        "High_Risk F1": 0.47
    },
    {
        "Approach": "SMOTE + XGBoost",
        "Accuracy": smote_results["Accuracy"],
        "Balanced Accuracy": smote_results["Balanced Accuracy"],
        "Macro F1": smote_results["Macro F1"],
        "High_Risk Recall": recall_score(
            y_val_encoded,
            y_val_pred_smote,
            labels=[1],
            average=None,
            zero_division=0
        )[0],
        "High_Risk F1": f1_score(
            y_val_encoded,
            y_val_pred_smote,
            labels=[1],
            average=None,
            zero_division=0
        )[0]
    }
])

comparison

,Approach,Accuracy,Balanced Accuracy,Macro F1,High_Risk Recall,High_Risk F1
0,Baseline XGBoost,0.953418,0.715139,0.735539,0.190000,0.290000
1,Balanced Sample Weight XGBoost,0.918000,0.867150,0.780906,0.780000,0.470000
2,SMOTE + XGBoost,0.950500,0.776712,0.788612,0.410293,0.453039


## Because the primary objective of the EMI eligibility system is to identify high-risk applicants and minimize the number of high-risk applicants incorrectly classified as lower-risk, High_Risk Recall is treated as the primary evaluation metric. Balanced XGBoost was therefore preferred over SMOTE + XGBoost because it achieved 78% High_Risk Recall compared with 41%

In [95]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()

y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

print("Classes:")
print(label_encoder.classes_)

Classes:
['Eligible' 'High_Risk' 'Not_Eligible']


In [96]:
from sklearn.utils.class_weight import compute_sample_weight

final_sample_weights = compute_sample_weight(
    class_weight="balanced",
    y=y_train_encoded
)

print("Sample weights shape:", final_sample_weights.shape)

Sample weights shape: (323840,)


In [97]:
from xgboost import XGBClassifier

final_xgb = XGBClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1,
    eval_metric="mlogloss"
)

In [98]:
final_xgb.fit(
    X_train_final,
    y_train_encoded,
    sample_weight=final_sample_weights
)

print("Final model trained successfully.")

Final model trained successfully.


In [99]:
y_test_pred = final_xgb.predict(
    X_test_final
)

In [100]:
from sklearn.metrics import (
    accuracy_score,
    balanced_accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

print("FINAL TEST RESULTS")
print("===================")

print("Accuracy:",
      accuracy_score(
          y_test_encoded,
          y_test_pred
      ))

print("Balanced Accuracy:",
      balanced_accuracy_score(
          y_test_encoded,
          y_test_pred
      ))

print("Macro Precision:",
      precision_score(
          y_test_encoded,
          y_test_pred,
          average="macro",
          zero_division=0
      ))

print("Macro Recall:",
      recall_score(
          y_test_encoded,
          y_test_pred,
          average="macro",
          zero_division=0
      ))

print("Macro F1:",
      f1_score(
          y_test_encoded,
          y_test_pred,
          average="macro",
          zero_division=0
      ))

FINAL TEST RESULTS
Accuracy: 0.9160326086956522
Balanced Accuracy: 0.8760996516403461
Macro Precision: 0.7564311062049535
Macro Recall: 0.8760996516403461
Macro F1: 0.7820593737257298


In [101]:
from sklearn.metrics import classification_report

print(
    classification_report(
        y_test_encoded,
        y_test_pred,
        target_names=label_encoder.classes_,
        zero_division=0
    )
)

              precision    recall  f1-score   support

    Eligible       0.94      0.89      0.92     14889
   High_Risk       0.33      0.81      0.47      3497
Not_Eligible       1.00      0.93      0.96     62574

    accuracy                           0.92     80960
   macro avg       0.76      0.88      0.78     80960
weighted avg       0.96      0.92      0.93     80960



# Balanced XGBoost achieved 87.61% balanced accuracy, 78.21% Macro F1, and 81% recall for the High_Risk class on the unseen test set. Since identifying High_Risk applicants was the primary business objective, the balanced XGBoost model was selected despite its lower overall accuracy compared with the unbalanced baseline.

In [102]:
import joblib

In [103]:
joblib.dump(
    final_xgb,
    "emi_xgboost_balanced.pkl"
)

print("XGBoost model saved successfully.")

XGBoost model saved successfully.


In [104]:
joblib.dump(
    label_encoder,
    "emi_label_encoder.pkl"
)

print("Label encoder saved successfully.")

Label encoder saved successfully.


In [105]:
selected_features = list(X_train_final.columns)

joblib.dump(
    selected_features,
    "emi_selected_features.pkl"
)

print("Selected features saved successfully.")
print("Number of selected features:", len(selected_features))

Selected features saved successfully.
Number of selected features: 15


In [106]:
import os

files = [
    "emi_xgboost_balanced.pkl",
    "emi_label_encoder.pkl",
    "emi_selected_features.pkl"
]

for file in files:
    if os.path.exists(file):
        print(f"✅ {file}")
    else:
        print(f"❌ {file}")

✅ emi_xgboost_balanced.pkl
✅ emi_label_encoder.pkl
✅ emi_selected_features.pkl


In [107]:
loaded_model = joblib.load(
    "emi_xgboost_balanced.pkl"
)

loaded_encoder = joblib.load(
    "emi_label_encoder.pkl"
)

loaded_features = joblib.load(
    "emi_selected_features.pkl"
)

print("Model loaded:", type(loaded_model))
print("Classes:", loaded_encoder.classes_)
print("Number of features:", len(loaded_features))

Model loaded: <class 'xgboost.sklearn.XGBClassifier'>
Classes: ['Eligible' 'High_Risk' 'Not_Eligible']
Number of features: 15


In [108]:
loaded_test_pred = loaded_model.predict(
    X_test_final
)

print(
    "Predictions identical:",
    (loaded_test_pred == y_test_pred).all()
)

Predictions identical: True


In [109]:
import zipfile

files = [
    "emi_xgboost_balanced.pkl",
    "emi_label_encoder.pkl",
    "emi_selected_features.pkl"
]

with zipfile.ZipFile(
    "emi_classification.zip",
    "w",
    zipfile.ZIP_DEFLATED
) as zipf:

    for file in files:
        zipf.write(file)

print("emi_classification.zip created successfully!")

emi_classification.zip created successfully!


In [110]:
with zipfile.ZipFile("emi_classification.zip", "r") as zipf:
    print("Files inside ZIP:")
    
    for file in zipf.namelist():
        print("✅", file)

Files inside ZIP:
✅ emi_xgboost_balanced.pkl
✅ emi_label_encoder.pkl
✅ emi_selected_features.pkl


In [111]:
import os

print("ZIP exists:", os.path.exists("emi_classification.zip"))
print(
    "ZIP size:",
    round(os.path.getsize("emi_classification.zip") / (1024 * 1024), 2),
    "MB"
)

ZIP exists: True
ZIP size: 0.87 MB


In [112]:
from IPython.display import FileLink

FileLink("emi_classification.zip")

/kaggle/working/emi_classification.zip